### PR TITLE
[P2 Bug] Fix sandstorm/hail interaction with semi invulnerable state

### LINKED ISSUE #4042: [Bug] Sandstorm and Hail damage Triggers on the Semi-Invulnerable Turn of Dig and Dive
### Describe the bug

Diving with either Hail or Sandstorm active still triggers damage on a pokemon in the semi-invulnerable turn of Dig or Dive. 

Link to Discord's Bug report: (Sandstorm hits pokemon using dive)
https://discord.com/channels/1125469663833370665/1246312758958624778

### Session export file

_No response_

### User data export file

_No response_

### Expected behavior

Bulba pages for Dive and Dig

> Sandstorm, hail, and Shadow Sky will not cause damage during the semi-invulnerable turn.

> During the semi-invulnerable turn, the user is also immune against damage from sandstorm and hail.

### Screenshots / Videos

Dive

https://github.com/user-attachments/assets/8853637d-af4d-4c53-b9a2-2016a1247e59

Dig

https://github.com/user-attachments/assets/ab8f99ce-0b05-4fd2-93b8-499fb93fc1aa

### What OS did you observe the bug on?

Other

### If other please specify

gameplay

### Which browser do you use?

Other

### If other please specify

gameplay

### Additional context

Pokemon in the semi-invulnerable turn of Bounce, Fly, Phantom and Shadow Force correctly take damage.

### PR DESCRIPTION
**Pull request by torranx**
## What are the changes the user will see?
A Pokemon in a semi-invulnerable state should no longer be damaged by Hail/Sandstorm

## Why am I making these changes?
closes #4042 

## What are the changes from a developer perspective?
add new condition on inflicting weather damage

## How to test the changes?
use Dive/Fly on sandstorm/hail weather
or
`npm run test weather_sandstorm`
`npm run test weather_hail`

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [ ] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [ ] Are the changes visual?
  - [ ] Have I provided screenshots/videos of the changes?

### CODE REVIEW

**Review by Tempo-anon — APPROVED** [review]
Code looks good to me. Just want to confirm (on showdown) if flying/diving/digging actually does prevent sand/hail damage

**Comment by Snailman11** [comment]
> Code looks good to me. Just want to confirm (on showdown) if flying/diving/digging actually does prevent sand/hail damage

https://replay.pokemonshowdown.com/gen9customgame-2195182905-dh4jembqzrgshlkf4oed9miqur0vmrupw

Smogon on Dive 
> On the first turn, the user avoids all attacks other than Surf and Whirlpool but takes double damage from them, and is also unaffected by weather.

Smogon on Dig
> On the first turn, the user avoids all attacks other than Earthquake and Magnitude but takes double damage from them, and is also unaffected by weather.

File: src/phases/weather-effect-phase.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import BattleScene from "#app/battle-scene.js";
 import { applyPreWeatherEffectAbAttrs, SuppressWeatherEffectAbAttr, PreWeatherDamageAbAttr, applyAbAttrs, BlockNonDirectDamageAbAttr, applyPostWeatherLapseAbAttrs, PostWeatherLapseAbAttr } from "#app/data/ability.js";
 import { CommonAnim } from "#app/data/battle-anims.js";
+import { SemiInvulnerableTag } from "#app/data/battler-tags.js";
```
Comment (flx-sta) [inline src/phases/weather-effect-phase.ts:4]: Groundhog day
```suggestion
import { SemiInvulnerableTag } from "#app/data/battler-tags";
```

**Review by flx-sta — CHANGES_REQUESTED** [review]

File: src/phases/weather-effect-phase.ts (hunk 1 above)
Comment (DayKev) [reply src/phases/weather-effect-phase.ts:4]: Remove all the `.js` while you're here, not just from the new import.

**Comment by Tempo-anon** [comment]
> It doesn't look like Fly should avoid Sand/Hail damage, only Dig and Dive.

I seemed to recall taking damage from weather while using fly. That also means Fly needs bespoke behavior compared to all the other semi invulnerable moves then :pikastare:

**Review by Tempo-anon — CHANGES_REQUESTED** [review]
Fly why

**Comment by flx-sta** [comment]
Back to `Draft` for Fly fix

**Comment by Snailman11** [comment]
***Additional context***
Pokemon in the semi-invulnerable turn of Bounce, Fly, Phantom and Shadow Force correctly take damage.

Not sure if this was seen?

**Review by Tempo-anon — APPROVED** [review]
You're telling me being underwater and underground will protect you from sand/hail but teleporting into the shadow dimension won't

**Review by CodeTappert — APPROVED** [review]

**Pull request merged**